### PR TITLE
Unpack tempfile's callables

### LINKED ISSUE #1000: Issue with open_in_editor
Seems like #851 didn't properly address tempfile/tempfile_suffix being dynamic/lambda.

https://github.com/prompt-toolkit/python-prompt-toolkit/blob/c8f5b5cd6048935c64ba451443af7962ad5d9cc2/prompt_toolkit/buffer.py#L1407-L1430

Currently, it's always ending up in the first clause since PromptSession passes the values as `lambda: self.tempfile`.

https://github.com/prompt-toolkit/python-prompt-toolkit/blob/c8f5b5cd6048935c64ba451443af7962ad5d9cc2/prompt_toolkit/shortcuts/prompt.py#L467-L487

### PR DESCRIPTION
**Pull request by njalerikson**
Need to unpack tempfile and tempfile_suffix callables before deciding whether creating a simple or advanced temp file for the editor.

Resolves #1000

### CODE REVIEW

File: prompt_toolkit/buffer.py [hunk 1]
```diff
@@ -1405,14 +1405,40 @@ def open_in_editor(self, validate_and_handle: bool = False) -> 'asyncio.Task[Non
             raise EditReadOnlyBuffer()
 
         # Write to temporary file
-        if self.tempfile:
+        try:
```
Comment (jonathanslenders) [inline prompt_toolkit/buffer.py]: This function is becoming long. Is there any way to break it up in maybe two or three functions. (It is hard to review).

File: prompt_toolkit/utils.py [hunk 2]
```diff
@@ -261,6 +262,14 @@ def take_using_weights(
         i += 1
 
 
+def to_type(value: Union[Callable[[], _T], _T]) -> _T:
```
Comment (jonathanslenders) [inline prompt_toolkit/utils.py]: Can you rename it to something like ``call_if_callable``. I think the naming is a little confusing.

Comment (njalerikson) [reply prompt_toolkit/utils.py]: Renamed

File: prompt_toolkit/buffer.py (hunk 1 above)
Comment (njalerikson) [reply prompt_toolkit/buffer.py]: Split tempfile creation into simple vs complex.

**Comment by jonathanslenders** [comment]
Thanks! Merged in: https://github.com/prompt-toolkit/python-prompt-toolkit/commit/c7eb02c50c384e6ab2c03225e5c76a2242a2752e

**Comment by jonathanslenders** [comment]
@njalerikson : Can you please have a look at this PR: https://github.com/prompt-toolkit/python-prompt-toolkit/pull/1007 
This contains a few improvements to this code.

There is also this line:
```python
 suffix = str(suffix) if suffix else None
```
for which I don't understand the purpose. Can you please explain what this is doing?

**Comment by njalerikson** [comment]
Ah I do like that change, much cleaner, this was one of the parts I wasn’t 100% happy with.

The suffix casting to string is likely unnecessary. It’s just catching any cases where the user passes some non-string (or non-None), I have no real opinion about it.

**Pull request closed without merging**